fixed crash and missing cards when one client deletes a pile while another adds a card

ArnoldSmith86 opened the pull request:
Fixes #2360.

This PR triggers conflict resolution when a client sets a `parent` to a widget that was - unbeknownst to it - already deleted by a different client.

It also adds `simulateServerLag` in `config.json`. When set to a number in milliseconds, it delays any messages to the clients. That makes it a lot easier to test conflict resolution in local environments.

---

PR-SERVER-BOT: You can play around with it here: https://test.virtualtabletop.io/PR-2363/pr-test (or any other room on that server)

bjalder26 commented:
This seems hard to trigger, is the code too involved to plug into jsfiddle with some fake variables and try a bunch of deltas in receiveDelta()?

ArnoldSmith86 commented:
Using my phone and my laptop, I can trigger this pretty easily.. I'm pretty sure I'll get this to work soon.

96LawDawg commented:
I could also trigger it.

ArnoldSmith86 commented:
I think the detection is working as intended but it also crashes if the conflicting deltas are the other way around. Gotta add another test for that.

ArnoldSmith86 commented:
Huh. The conflict detection seems to work now but the client still has to handle the broken delta. It goes like this:

1. Client A sends delta 1 that sets any parent to `widgetA`.
2. The server receives delta 1, accepts it and broadcasts it.
3. Client B didn't receive delta 1 yet and sends delta 2 that removes `widgetA`.
4. Client B receives delta 1 and crashes because it can't set parent to `widgetA` anymore because it deleted it. :boom: 
5. The server receives delta 2, notices a conflict and sends the game state to client B. But client B already crashed.

ArnoldSmith86 commented:
It no longer crashes but when the events of the last comment happen, there are ghost widgets remaining. Gotta do some deeper fixing.

ArnoldSmith86 commented:
I think this is working now.

ArnoldSmith86 commented on client/js/serverstate.js:
```diff
@@ -430,7 +430,8 @@ function receiveStateFromServer(args) {
   deltaID = args._meta.deltaID;
-  for(const widget of widgetFilter(w=>w.get('parent')===null))
+  const topSurface = $('#topSurface');
+  for(const widget of widgetFilter(w=>w.domElement.parentElement === topSurface))
```
Each one of those removes its children recursively. IIRC, the leaves of the tree have to be removed first to avoid issues with already removed parents.

rogerl50 replied:
Sorry, I wasn't clear. How do you know that the widgets removed are on the top surface because their original parent does not exist? That is, why doesn't this just clear the table?

ArnoldSmith86 replied:
I'm still not quite following. This _is_ clearing the table because the server sent a complete new state.

96LawDawg commented:
In testing, there is a strange visual glitch.  A card is displayed for about 1/2 second at 0,0 in the room, then it appears in the holder.  It is very noticeable.

ArnoldSmith86 commented:
Yeah, it tries to set its parent to a non-existant widget. I guess I could hide the widget.

ArnoldSmith86 commented:
I added a class `limbo` to the affected widgets now that hides them from play but makes them look like this in the editor:

![image](https://github.com/user-attachments/assets/7b46b1e1-955e-42db-a5d9-4188b4912d8e)

It can simply be edited like any other widget so you can correct the invalid parent. _But_ at least in the context of this PR, this should basically only occur when a conflict is triggered and the room reloaded anyway so you shouldn't really have time to interact with it.

This does _not_ change the behavior that widgets with an invalid parent aren't even added to the room when a new state is loaded. If we decide to do that, I would apply the same `limbo` class but that's a change for another PR.

Note that the class simply sets `opacity` and `border`. That might cause visual glitches in other widgets.

---

I also added a new `config.json` option: `simulateServerLag`. When set to a number in milliseconds, it delays any messages to the clients. That makes it a lot easier to test conflict resolution in local environments.

ArnoldSmith86 commented:
@96LawDawg Are you sure that this fixes #27? Sounds unrelated. Maybe it's fixed in production already? Would be a nice use-case for the new `simulateServerLag`.

96LawDawg commented:
> @96LawDawg Are you sure that this fixes #27? Sounds unrelated. Maybe it's fixed in production already? Would be a nice use-case for the new `simulateServerLag`.

Of course not.  It just seemed like the symptoms were the same from what I remember.  If you aren't sure, then de-link it.

RaphaelAlvez reviewed (dismissed):
Do we really need a border in limbo?

ArnoldSmith86 commented:
> Do we really need a border in limbo?

No. I'm trying to fix a very important bug here. Change the limbo CSS to whatever you want. 🤣

The discussion in Discord lets me believe that we'll allow invalid parents soonish and use "limbo" for that. Let's just style and/or rename "limbo" when we do that.

96LawDawg commented:
Is this ready for final testing or is there still something to do?

ArnoldSmith86 commented:
It's ready. I hope. 😆

96LawDawg reviewed (approved):
Approved through test.  I was able to generate a limbo card manually in the dev console to see what it looks like.  I was able to generate it through the automation set up in the test room.  Previously everything had worked and it still does.  There is some flashing, but not the jarring display of the card at 0,0.  In edit mode, you can briefly see the limbo card at 0,0.

As is, this is a big improvement and does what it is supposed to.  Ideally, there would not be any flashing to indicate something is amiss, but that is just a dream.  This will help prevent lost cards for sure.
